### PR TITLE
Fix string to double conversion and row equivalent comparison

### LINKED ISSUE #5225: [BUG] discrepency in casting String to double
**Describe the bug**
When casting '1.7976931348623159E308' to double. CPP returns [`Inf`](https://wandbox.org/permlink/a2gkgDL8UriLUFGn) but cudf returns `1.7976931348623151E308` 

**Steps/Code to reproduce bug**
Create a string column with '1.7976931348623159E308' and cast it to double. 

**Expected behavior**
Should return infinity like [CPP](https://wandbox.org/permlink/a2gkgDL8UriLUFGn).

### LINKED ISSUE #5731: [BUG] expect_columns_equivalent handles infinity incorrectly.
**Describe the bug**
When comparing any floating point value with (+/-) infinity, `expect_columns_equivalent` indicates the values are equivalent.

**Steps/Code to reproduce bug**
```
auto inf = std::numeric_limits<double>::infinity();
auto a = fixed_width_column_wrapper<double>({inf, inf, inf, inf});
auto b = fixed_width_column_wrapper<double>({0, 1, -1, -inf});
expect_columns_equivalent(a, b); // passes
```

**Expected behavior**
Any of these should fail:
```
auto inf = std::numeric_limits<double>::infinity();
auto inf_col =   cudf::test::fixed_width_column_wrapper<double>({inf});
expect_columns_equivalent(inf_col, fixed_width_column_wrapper<double>({0}));
expect_columns_equivalent(inf_col, fixed_width_column_wrapper<double>({1}));
expect_columns_equivalent(inf_col, fixed_width_column_wrapper<double>({-1}));
expect_columns_equivalent(inf_col, fixed_width_column_wrapper<double>({-inf}));
```
All of these should pass:
```
auto inf = std::numeric_limits<double>::infinity();
auto inf_col =   cudf::test::fixed_width_column_wrapper<double>({inf});
auto neg_inf_col =   cudf::test::fixed_width_column_wrapper<double>({-inf});
expect_columns_equivalent(inf_col, fixed_width_column_wrapper<double>({inf}));
expect_columns_equivalent(neg_inf_col, fixed_width_column_wrapper<double>({-inf}));
```

**Additional Context**
`expect_columns_equal` behaves as expected.

### PR DESCRIPTION
**Pull request by ttnghia**
This PR fixes #5225 and fixes #5731 along with several improvements. In particular:

  * Fixes the function `stod` in `convert_floats.cu`, which incorrectly uses `max_mantissa`, thus produces incorrect results as mentioned in the issue #5225. Note that this PR still cannot produce perfectly matched results with `std::atof`, since handling float values near `inf` is very difficult.
  * Fixes the class `corresponding_rows_not_equivalent` which didn't handle `inf` and `nan`, thus it returned "equivalent" when comparing a valid float number with `inf`.
  * Adds a test case for those fixes.
  * Rewrite test cases for string=>float coversion to use `std::atof` results for comparison.

### CODE REVIEW

File: cpp/tests/strings/floats_tests.cpp [hunk 1]
```diff
@@ -128,7 +128,8 @@ TEST_F(StringsConvertTest, ToFloats64)
                                      "456e",
                                      "-1.78e+5",
                                      "-122.33644782",
-                                     "12e+309"};
+                                     "12e+309",
+                                     "1.7976931348623159E308"};
```
Comment (davidwendt) [inline cpp/tests/strings/floats_tests.cpp:132]: I worked on this a bit last year but was not able to get consistent results for the following test values compared to `std::stod`
```
"1.7976931348623158E308"
"1.79769313486231581E308"
"-1.7976931348623158E308"
"-1.79769313486231581E308"
```
I would be curious what results you get for these and if you could add them to the gtest.

**Review by davidwendt — CHANGES_REQUESTED** [review]
This is excellent. 
Looks like you reformatted the CMakeLists.txt file.
You should change that back please.

File: cpp/tests/utilities/column_utilities.cu [hunk 2]
```diff
@@ -127,11 +127,20 @@ class corresponding_rows_not_equivalent {
       column_device_view const& lhs, column_device_view const& rhs, size_type index)
     {
       if (lhs.is_valid(index) and rhs.is_valid(index)) {
-        int ulp = 4;  // value taken from google test
-        T x     = lhs.element<T>(index);
-        T y     = rhs.element<T>(index);
-        return std::abs(x - y) > std::numeric_limits<T>::epsilon() * std::abs(x + y) * ulp &&
-               std::abs(x - y) >= std::numeric_limits<T>::min();
+        const T x = lhs.element<T>(index);
+        const T y = rhs.element<T>(index);
```
Comment (jrhemstad) [inline cpp/tests/utilities/column_utilities.cu:131]: West Const rules the land here :) 
```suggestion
        T const x = lhs.element<T>(index);
        T const y = rhs.element<T>(index);
```

File: cpp/tests/strings/floats_tests.cpp (hunk 1 above)
Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:132]: Humm, it seems that my fix still doesn't work right for your examples. I'm looking at the code again.

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:132]: It's weird. I ran a tests with C++ `std::stod`. "1.7976931348623158E308" is converted to `1.7976931348623157081452742e+308`. The maximum allowed double number is also `1.7976931348623157081452742e+308` (the same), which is larger than the number given in the string.

So, somehow, `std::stod` also incorrectly converts string to double.

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:132]: Alright, my fix still produces discrepant results for your examples. The root cause of that is the line:
```
static_cast<FloatType>(digits) * exp10(static_cast<FloatType>(exp_ten))
```
Multiplying two very large floating point numbers leads to the problem. Take the first 2 strings as examples:
 * For the first string ("1.7976931348623158E308"), parsing it we will have `digits = 17976931348623158` and `exp_ten = 292`. Multiplying those number will produce `inf`, which is our result. `std::stod` produces `1.7976931348623157081452742e+308`.

 * For the second string ("1.79769313486231581E308"), parsing it we will have `digits = 179769313486231581` and `exp_ten = 291`. Multiplying those number will produce `1.7976931348623157081452742e+308
` --- equals to the value of `std::numeric_limits<double>::max()`, which is our result. `std::stod` produces `inf`.

I have no idea how to overcome those.


I have no solution for this yet.

Comment (davidwendt) [reply cpp/tests/strings/floats_tests.cpp:132]: Yes. I kind of gave up on this for awhile since I'm not sure if it completely satisfies #5225 
I think your solution is much better than the current implementation though.
Could check with the Spark team whether or not this solution works for them?

**Review by kkraus14 — APPROVED** [review]
cmake lgtm

File: cpp/tests/strings/floats_tests.cpp (hunk 1 above)
Comment (revans2) [reply cpp/tests/strings/floats_tests.cpp:132]: I am now seeing some issues with float parsing. The cudf java unit tests are failing when it tries to parse "3.4028236E38" to a float.  CUDF now returns "3.4028234663852886E38" as the value, but before it would return `Inf` and when I check both java and atof (casting the result to a float from a double) I also get `Inf`.

Comment (revans2) [reply cpp/tests/strings/floats_tests.cpp:132]: Just be aware that there are also some bugs in the java test for parsing doubles.  Looks like when the test was written someone copied and pasted in the wrong value for max and min double values verifying that the code is wrong. So I will put up a PR to fix that test as soon as this is working too.

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:132]: Can you pull my latest commit and test again please?

Comment (revans2) [reply cpp/tests/strings/floats_tests.cpp:118]: I upmerged and rebuild everything and I am still getting failures.  The C++ tests fail for me with 

```
[ RUN      ] StringsConvertTest.ToFloats32
../tests/utilities/column_utilities.cu:234: Failure
Failed
differences:
lhs[17] = 3.40282326e+38, rhs[17] = inf

Google Test trace:
../tests/strings/floats_tests.cpp:67:  <--  line of failure

[  FAILED  ] StringsConvertTest.ToFloats32 (0 ms)
```

and the java tests also fail with `"3.4028236E38"` when parsed to a float is not `inf` but is `3.4028232635611926E38` and `"1.7976931348623158E308"` when parsed to a double is not `inf` but is `1.7976931348623157E308`.

Comment (davidwendt) [reply cpp/tests/strings/floats_tests.cpp:118]: I tried even more decimal places and did not find a discernible pattern from `std:stod`. 
I don't know you will ever get agreement on floating point values here across C++, CUDA, Java, and Python libraries.

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:118]: I just ran my tests (`float_tests.cpp`) and all tests pass! I'm `make clean` and `make -j` to check again...

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:118]: Fresh build and test. All test passed! "3.4028236E38" is definitely on the list of string=>float32 test.

Boby, can you check if you have pulled my latest commit? Thanks.

Comment (revans2) [reply cpp/tests/strings/floats_tests.cpp:118]: I am still seeing errors. I just cleaned and rebuilt everything so I am not sure why it is different.

Comment (revans2) [reply cpp/tests/strings/floats_tests.cpp:118]: I should point out that CI is showing similar failures on some platforms.

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:118]: It seem that the old conversion way (always convert string to double then convert to float) works better as double-precision can handle edge float32 values (near inf) better. I change back my code to use that way, instead of converting string directly to float32. Now all CI tests pass. Boby, I believe that your tests would pass too.

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:118]: Note that parsing string to floats involves some multiplication of the float values. Thus, it's extremely difficult to handle values near inf as for those operations overflow and round-off can happen anytime and lead to invalid results.

Comment (revans2) [reply cpp/tests/strings/floats_tests.cpp:118]: The latest changes fixed all of my issues

**Review by revans2 — APPROVED** [review]
From my perspective this fixes all of the issues that I was seeing with accuracy.

File: cpp/tests/strings/floats_tests.cpp (hunk 1 above)
Comment (revans2) [reply cpp/tests/strings/floats_tests.cpp:118]: Sorry I updated the tests to check other values, and I forgot to update them back. I am still getting issues for double parsing of `"1.7976931348623158E308"`  This is still better then what was before. But it is not 100% perfect.

Comment (revans2) [reply cpp/tests/strings/floats_tests.cpp:118]: I should say that it is good enough that I am fine with checking it in. It is a corner case that we can document, but it would be nice to fix it.

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:118]: Thanks Boby!!!

I don't have problem with "1.7976931348623158E308" on my computer---forever excuse 😄.

Comment (davidwendt) [reply cpp/tests/strings/floats_tests.cpp:118]: Ok, so how is this now different than the original code? 
You are hardcoding `stof<double>` so the template parameter is no longer required.
If this is working better than the change is likely minor.

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:118]: Yes, the change is insignificant. I just tri`ed to improve the original function `atof` by rewriting it a little bit, avoiding round-off on the integer mantissa before converting it to a double number. I also tried to reduce the round-off by moving the exponent around. Again, like @davidwendt did, I found it extremely difficult to correctly handle edge values. 

I also dig into the code of `std::atof` from various `libc` sources---they are huge, and are much more complex, with many macros to handle different platforms and edge cases. So, I  think that I/we may never be able to accurately port `std::atof` to CUDA. Our implementation is good enough.

Comment (jrhemstad) [reply cpp/tests/strings/floats_tests.cpp:118]: > I also dig into the code of std::atof from various libc sources

Be very careful with looking at any libc implementations. Most of them are GPL and you cannot use anything you find there. It'd be best to avoid looking at them all together to avoid the possibility or appearance of impropriety.

Comment (davidwendt) [reply cpp/tests/strings/floats_tests.cpp:118]: > Although my template function is now only used for instantiating double, I prefer to keep it as a template function.

I disagree with this approach. I would rather improve on the magic numbers usage.

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:118]: Ah sure, thanks for let me know that.

Comment (ttnghia) [reply cpp/tests/strings/floats_tests.cpp:118]: Ah right. OK, I've changed that---the template function `stof` is now again `stod`.

File: cpp/src/strings/convert/convert_floats.cu [hunk 3]
```diff
@@ -132,8 +140,6 @@ struct string_to_float_fn {
   __device__ FloatType operator()(size_type idx)
   {
     if (strings_column.is_null(idx)) return static_cast<FloatType>(0);
-    // the cast to FloatType will create predictable results
-    // for floats that are larger than the FloatType can hold
     return static_cast<FloatType>(stod(strings_column.element<string_view>(idx)));
```
Comment (davidwendt) [inline cpp/src/strings/convert/convert_floats.cu:143]: Recommend re-instating this comment or a rephrasing of it perhaps since this was a vital part of the code behavior.

File: cpp/benchmarks/string/convert_floats_benchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  }
+  return numbers;
+}
+
+template <class FloatType>
+static const std::vector<std::string>& get_floats_numbers_as_string(int64_t array_size)
+{
+  static std::unordered_map<int64_t, std::vector<std::string>> str_arrays;
+  auto& numbers_str = str_arrays[array_size];
+  if (numbers_str.size() == 0) {
```
Comment (davidwendt) [inline cpp/benchmarks/string/convert_floats_benchmark.cpp:55]: Is this map necessary? Are the vectors re-used in the gbenchmarks?

Comment (ttnghia) [reply cpp/benchmarks/string/convert_floats_benchmark.cpp:55]: Yes, the numbers/strings vectors are reused for the case of float32/float64 conversion benchmarks.

File: cpp/tests/utilities/column_utilities.cu [hunk 5]
```diff
@@ -127,11 +127,20 @@ class corresponding_rows_not_equivalent {
...
-        return std::abs(x - y) > std::numeric_limits<T>::epsilon() * std::abs(x + y) * ulp &&
-               std::abs(x - y) >= std::numeric_limits<T>::min();
+        T const x = lhs.element<T>(index);
+        T const y = rhs.element<T>(index);
+
+        // Must handle inf and nan separately
+        if (std::isinf(x) || std::isinf(y)) {
+          return x != y;  // comparison of (inf==inf) returns true
+        } else if (std::isnan(x) || std::isnan(y)) {
+          return std::isnan(x) ^ std::isnan(y);  // comparison of (nan==nan) returns false
```
Comment (davidwendt) [inline cpp/tests/utilities/column_utilities.cu:137]:
```suggestion
          return std::isnan(x) != std::isnan(y);  // comparison of (nan==nan) returns false
```
We know one of them is already `true`

File: cpp/benchmarks/string/convert_floats_benchmark.cpp (hunk 4 above)
Comment (ttnghia) [reply cpp/benchmarks/string/convert_floats_benchmark.cpp:55]: Maybe not in this benchmark. I observed that for each size the number/string vector is used only once.

Comment (davidwendt) [reply cpp/benchmarks/string/convert_floats_benchmark.cpp:55]: I see the case for the number vector since it is used for both the to/from benchmarks. But the string one is used only in the from-string benchmark. So I think this one can be simplified.

File: cpp/benchmarks/string/convert_floats_benchmark.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "../fixture/benchmark_fixture.hpp"
+#include "../synchronization/synchronization.hpp"
```
Comment (davidwendt) [inline cpp/benchmarks/string/convert_floats_benchmark.cpp:18]:
```suggestion
#include <fixture/benchmark_fixture.hpp>
#include <synchronization/synchronization.hpp>
```

**Review by davidwendt — CHANGES_REQUESTED** [review]

**Review by davidwendt — APPROVED** [review]
Looks good. Thanks for doing this.

**Comment by ttnghia** [comment]
@gpucibot merge.

File: cpp/src/strings/convert/convert_floats.cu [hunk 7]
```diff
@@ -58,15 +58,15 @@ __device__ inline double stod(string_view const& d_str)
...
+  double sign{1.0};
   if (*in_ptr == '-' || *in_ptr == '+') {
     sign = (*in_ptr == '-' ? -1 : 1);
     ++in_ptr;
   }
-  unsigned long max_mantissa = 0x0FFFFFFFFFFFFF;
-  unsigned long digits       = 0;
-  int exp_off                = 0;
-  bool decimal               = false;
+  constexpr uint64_t max_holding = (std::numeric_limits<uint64_t>::max() - 9L) / 10L;
```
Comment (vuule) [inline cpp/src/strings/convert/convert_floats.cu:66]: would be nice to have a comment on why this value is used

File: cpp/benchmarks/string/convert_floats_benchmark.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  state.SetBytesProcessed(state.iterations() * floats_size);
+}
+
+#define CV_TO_FLOATS_BENCHMARK_DEFINE(name, float_type_id)                  \
+  BENCHMARK_DEFINE_F(StringToFloatNumber, name)(::benchmark::State & state) \
+  {                                                                         \
+    convert_to_float_number<float_type_id>(state);                          \
+  }                                                                         \
+  BENCHMARK_REGISTER_F(StringToFloatNumber, name)                           \
+    ->RangeMultiplier(1 << 2)                                               \
```
Comment (vuule) [inline cpp/benchmarks/string/convert_floats_benchmark.cpp:111]: just a suggestion, '4' seems clearer to me
```suggestion
    ->RangeMultiplier(4)                                                    \
```

File: cpp/benchmarks/string/convert_floats_benchmark.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    ->UseManualTime()                                                       \
+    ->Unit(benchmark::kMicrosecond);
+
+#define CV_FROM_FLOATS_BENCHMARK_DEFINE(name, float_type)                     \
+  BENCHMARK_DEFINE_F(StringFromFloatNumber, name)(::benchmark::State & state) \
+  {                                                                           \
+    convert_from_float_number<float_type>(state);                             \
+  }                                                                           \
+  BENCHMARK_REGISTER_F(StringFromFloatNumber, name)                           \
+    ->RangeMultiplier(1 << 2)                                                 \
```
Comment (vuule) [inline cpp/benchmarks/string/convert_floats_benchmark.cpp:122]: same here
```suggestion
    ->RangeMultiplier(4)                                                      \
```

File: cpp/benchmarks/string/convert_floats_benchmark.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    cudf::test::UniformRandomGenerator<FloatType> rand_gen(std::numeric_limits<FloatType>::min(),
+                                                           std::numeric_limits<FloatType>::max());
+    std::generate_n(
+      std::back_inserter(numbers), array_size, [&rand_gen]() { return rand_gen.generate(); });
+  }
+  return numbers;
+}
+
+template <class FloatType>
+static std::vector<std::string> get_floats_numbers_as_string(int64_t array_size)
```
Comment (vuule) [inline cpp/benchmarks/string/convert_floats_benchmark.cpp:53]: Would it be faster to to this conversion on the GPU (unless there's a concern about using `cudf::strings::from_floats` here)?

File: cpp/benchmarks/string/convert_floats_benchmark.cpp [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+#include <numeric>
+#include <random>
+#include <unordered_map>
+
+namespace {
+
+// For each array_size, this function is called twice from both StringToFloatNumber and
+// StringFromFloatNumber classes. Thus, the results are cached for reuse.
+template <class FloatType>
+static const std::vector<FloatType>& get_float_numbers(int64_t array_size)
```
Comment (vuule) [inline cpp/benchmarks/string/convert_floats_benchmark.cpp:38]: Assuming you do the `to_string` conversion on the GPU, I would suggest using `create_random_table` here. The default distribution is Gaussian, which is arguably a more common for real data, and does affect this benchmark. You can still memoize the result.

**Review by vuule — CHANGES_REQUESTED** [review]
Couple suggestions, mostly related to data generation.

File: cpp/benchmarks/string/convert_floats_benchmark.cpp (hunk 10 above)
Comment (ttnghia) [reply cpp/benchmarks/string/convert_floats_benchmark.cpp:53]: Thanks for the recommendation. I've changed to use `from_floats`.

File: cpp/benchmarks/string/convert_floats_benchmark.cpp (hunk 11 above)
Comment (ttnghia) [reply cpp/benchmarks/string/convert_floats_benchmark.cpp:38]: Good idea. I've changed that.

Comment (ttnghia) [reply cpp/benchmarks/string/convert_floats_benchmark.cpp:38]: I just realized that we cannot use static variable to cache the random table: https://app.slack.com/client/T04SYRAP3/CDTANRCTT/thread/CDTANRCTT-1614629965.099000

So, just generate a random table each time we need it.

**Review by vuule — APPROVED** [review]

**Comment by ttnghia** [comment]
@gpucibot merge.

**Comment by harrism** [comment]
@ttnghia you can't merge if CI is not passing...

**Comment by harrism** [comment]
@gpucibot rerun tests

**Comment by ttnghia** [comment]
I see. Thanks Mark.
It seems that there are some issues with CI accounts so the server could not run.

**Comment by ttnghia** [comment]
@gpucibot merge.

**Comment by ttnghia** [comment]
@gpucibot merge.

**Comment by kkraus14** [comment]
@gpucibot merge
